### DISCUSSION TITLE
WiOn 50055 WiFi Plug - HLW8012 voltage divider

### DISCUSSION THREAD

**Discussion by JeffWDH** [arendst/Tasmota#20595] [Support request]
Hello,

I have a WiOn 50055 WiFi Plug that contains a HLW8012 power monitoring IC, however in the stock form there is a separate undocumented IC in between it and the ESP8266. I have bypassed the IC and wired the HLW8012 up like this:

GPIO4 - HLWBL CF1
GPIO5 - HLW8012 CF
GPIO16 - HLWBL SEL

Using this, I get accurate (compared to a Sonoff S31) current readings however the voltage reading increases when the current increases.

Examples:
![image](https://github.com/arendst/Tasmota/assets/13274150/a86dd182-d980-4420-8bbb-4817856efab1)

![image](https://github.com/arendst/Tasmota/assets/13274150/b2b4eaee-3704-43b8-83eb-042d8f1f0464)

As I understand it, the HLW8012 support was put in place based on the voltage divider recommended by the chip vendor and what is in use by the Sonoff POW R1:
1k, 6x470k
![hlw8012-circuit](https://github.com/arendst/Tasmota/assets/13274150/2bcc9064-675b-438a-9030-bbecb03e6e15)

However, this is not what is in the Wion 50055:
1k, 3x470k/1x412k 
![2020_0101_000334_001](https://github.com/arendst/Tasmota/assets/13274150/ac69b594-603c-4d39-a527-ad354199e777)

I've looked at the Tasmota source and I have been unable to figure out how to adjust the pre existing values to match this device. 

xnrg_01_hlw8012.ino has these values:
```
// Energy model type 0 (GPIO_HLW_CF) - HLW8012 based (Sonoff Pow, KMC70011, HuaFan, AplicWDP303075)
#define HLW_PREF            10000    // 1000.0W
#define HLW_UREF             2200    // 220.0V
#define HLW_IREF             4545    // 4.545A
```

How do those line up with the voltage divider? Does anybody have any suggestions?

Thank you!

FYI, I've documented the process of installing Tasmota on this device here:  
https://www.tarball.ca/posts/woods_wion_50055_wifi_plug_hacking/

This includes some high resolution images and pinouts of the board for reference.

**Comment by barbudor** [comment]
Hello
Have you read the section on Power Calibration ? https://tasmota.github.io/docs/Power-Monitoring-Calibration/
Command like VoltageSet should allow tasmota to automatically calculate the adjutsment factor

**Comment by JeffWDH** [comment]
I have used PowerSet, VoltageSet and CurrentSet commands to match the respective values from the S31. The current scales appropriately but the voltage increases/decreases when the current changes. You can see an example above with the side by side Wion/Sonoff S31 screenshots. The only difference is the first screenshot is from the "fan only" setting of a portable space heater and the second is with the heating element turned on.

**Reply by barbudor** [reply]
Ok, it was clear from your previous statement

Apart from the resistor values that are different, have you checked the rest of the cabling ?
Like an swapping of V1P and V1N that would confuse the chip

What is your load ? Is it purely resistive ?

Was the plug working as expected with the original firware ?
Any possibility that you have damaged something when removing the unwanted part ? (probably something like a MCU, Tuya or not)

**Comment by JeffWDH** [comment]
I enabled debug mode by uncommenting "#define HLW_DEBUG" and see these at loglevel 3:

```
15:58:51.800 NRG: power 205054, ui 1, cnt 10, smpl 32073 32097 32098 32133 32168 32207 32205 32179 32116 32116, sum 321392, mean 32139, median 32133
15:58:52.800 NRG: power 205218, ui 0, cnt 10, smpl 32091 32091 32091 32127 32125 32090 32090 32092 32130 32093, sum 321020, mean 32102, median 32092
15:58:53.800 NRG: power 205167, ui 1, cnt 10, smpl 32132 32131 32095 32131 32059 32132 32099 32131 32133 32133, sum 321176, mean 32117, median 32131
15:58:54.799 NRG: power 204795, ui 0, cnt 10, smpl 32132 32131 32133 32169 32131 32167 32166 32166 32204 32129, sum 321528, mean 32152, median 32166
15:58:55.800 NRG: power 205031, ui 1, cnt 10, smpl 32100 32137 32135 32170 32135 32170 32136 32136 32208 32172, sum 321499, mean 32149, median 32137
15:58:56.800 NRG: power 204947, ui 0, cnt 10, smpl 32130 32133 32096 32096 32131 32096 32095 32096 32132 32096, sum 321101, mean 32110, median 32096
```

Does this indicate that maybe my SEL output isn't working? I'm new to this code but my suspicion is the ui flag is flipping between voltage/current measurements.

**Reply by barbudor** [reply]
One possiblity could be that SEL is not connected properly and the HLW always present the current measure whatever tasmota believe it asking

**Reply by barbudor** [reply]
We came to the same conclusion simultaneously

**Reply by JeffWDH** [reply]
Can confirm SEL is indeed going high/low - stuck a LED to the SEL pin of the HLW8012 and it is blinking away. Not sure if it's supposed to but it seems to blink faster when there is a load.

**Reply by barbudor** [reply]
Looking quickly to the code, I feel it supposed to change state every 1.6 seconds : 8 ticks of 200ms timer
But from your log, I see it change every second

Do you see the period of the above logs to change in rythm with the led when the current increase ?

**Reply by JeffWDH** [reply]
I added this to troubleshoot:
```
Top of xnrg_01_hlw8012.ino:
float lastmillis = millis();

Line 150:
  if (PinUsed(GPIO_NRG_CF1)) {
    Hlw.cf1_timer++;
    if (Hlw.cf1_timer >= 8) {
      Hlw.cf1_timer = 0;
      Hlw.select_ui_flag = (Hlw.select_ui_flag) ? false : true;
      DigitalWrite(GPIO_NRG_SEL, 0, Hlw.select_ui_flag);
      AddLog(LOG_LEVEL_DEBUG, PSTR("SEL: %d millis"), (millis() - lastmillis));       # <-- Added this
      lastmillis = millis();                                                                                            # <-- and this
```

Relay on with no load:
```
18:20:17.000 SEL: 1598 millis
18:20:19.602 SEL: 1598 millis
18:20:21.200 SEL: 1597 millis
18:20:22.800 SEL: 1597 millis
18:20:24.401 SEL: 1597 millis
18:20:25.000 SEL: 1598 millis
18:20:27.601 SEL: 1598 millis
18:20:29.200 SEL: 1597 millis
18:20:30.800 SEL: 1598 millis
18:20:32.400 SEL: 1598 millis
18:20:33.000 SEL: 1598 millis
18:20:35.600 SEL: 1598 millis
18:20:37.200 SEL: 1598 millis
18:20:38.800 SEL: 1597 millis
18:20:40.400 SEL: 1597 millis
18:20:41.000 SEL: 1599 millis
18:20:43.600 SEL: 1599 millis
18:20:45.200 SEL: 1599 millis
18:20:46.800 SEL: 1598 millis
18:20:48.400 SEL: 1598 millis
18:20:49.000 SEL: 1598 millis
```

Relay on with 24 watt load:
```
18:21:33.600 SEL: 998 millis
18:21:34.600 SEL: 998 millis
18:21:35.600 SEL: 998 millis
18:21:36.600 SEL: 998 millis
18:21:37.601 SEL: 998 millis
18:21:38.600 SEL: 998 millis
18:21:39.600 SEL: 997 millis
18:21:40.600 SEL: 998 millis
18:21:41.600 SEL: 998 millis
18:21:42.600 SEL: 998 millis
18:21:43.601 SEL: 997 millis
18:21:44.600 SEL: 998 millis
18:21:45.600 SEL: 998 millis
18:21:46.600 SEL: 998 millis
18:21:47.600 SEL: 998 millis
18:21:48.600 SEL: 997 millis
18:21:49.601 SEL: 998 millis
18:21:50.601 SEL: 998 millis
18:21:51.600 SEL: 998 millis
18:21:52.601 SEL: 998 millis
```

Relay on with ~800W load
```
8:22:21.798 SEL: 798 millis
18:22:22.600 SEL: 797 millis
18:22:23.400 SEL: 797 millis
18:22:24.203 SEL: 801 millis
18:22:24.000 SEL: 795 millis
18:22:25.798 SEL: 798 millis
18:22:26.600 SEL: 798 millis
18:22:27.401 SEL: 797 millis
18:22:28.200 SEL: 798 millis
18:22:28.000 SEL: 798 millis
18:22:29.798 SEL: 798 millis
18:22:30.600 SEL: 798 millis
18:22:31.401 SEL: 797 millis
18:22:32.201 SEL: 799 millis
18:22:32.000 SEL: 796 millis
18:22:33.798 SEL: 798 millis
18:22:34.600 SEL: 798 millis
18:22:35.401 SEL: 798 millis
18:22:36.200 SEL: 798 millis
18:22:36.000 SEL: 798 millis
18:22:37.798 SEL: 797 millis
18:22:38.600 SEL: 798 millis
18:22:39.400 SEL: 798 millis
18:22:40.202 SEL: 798 millis
18:22:40.001 SEL: 797 millis
```

Although, now looking a bit at the code it seems like the measurement time can vary as I see some comments that indicate that:
"// Allow for 300 mSec set-up time and measure for up to 1 second"
"// We need up to HLW_SAMPLE_COUNT samples within 1 second (low current could take up to 0.3 second)"

Either way, the LED shows that SEL is connected and cycling and CF1 is connected as I am getting current readings. It's weird that the voltage is so wrong.

**Comment by JeffWDH** [comment]
Update: This is not a Tasmota specific issue... ESPHome seems to have the same voltage reading issue.

**Reply by JeffWDH** [reply]
https://community.home-assistant.io/t/decoding-power-bar-with-energy-monitoring/622841/8

This post suggests that 3.3v from the esp8266 may not be high enough voltage to be a high signal for the HLW8012 SEL pin. I'll have to figure out what my options are and see if I have any level shifters laying around.

**Reply by barbudor** [reply]
That may explain why there is a GPIO for SEL_Inverted in case a transistor is used to shift up the signal. 
If you have a spare NPN or Mosfet-N plus a couple of resistors it should be straight forward.

**Comment by JeffWDH** [comment]
Level shifting did the trick - had to use inverted SEL but it's working as expected now after a quick calibration.

Thanks for talking this through @barbudor!
